File: Python/1238 - Silver Cow Party.py
import heapq
import sys
input = sys.stdin.readline


def adj_check(u, v, w):
    global adj
    adj[u].append((w, v))


def dijkstra(start, end):
    result = [INF for _ in range(n + 1)]

    result[start] = 0
    heapq.heappush(heap, (0, start))

    while heap:
        cur_w, cur_node = heapq.heappop(heap)

        if result[cur_node] < cur_w:
            continue

        for edge in adj[cur_node]:
            next_w, next_node = edge
            next_w += cur_w

            if next_w < result[next_node]:
                result[next_node] = next_w
                heapq.heappush(heap, (next_w, next_node))

    return result[end]


if __name__ == '__main__':
    INF = int(1e9)
    heap = []
    adj = {}
    ans = -1

    n, m, x = map(int, input().split())

    for i in range(n+1):
        adj[i] = []

    for _ in range(m):
        node1, node2, weight = map(int, input().split())
        adj_check(node1, node2, weight)

    for i in range(n):
        temp = dijkstra(i+1, x) + dijkstra(x, i+1)
        ans = max(ans, temp)

    print(ans)

File: C++/1238 - Silver Cow Party.cpp
#include<iostream>
#include<string.h>
#include<vector>
#include<queue>
#include<algorithm>
using namespace std;


typedef pair<int, int> pii;
const int INF = 1e9 + 1;
const int MAX_N = 1e3 + 1;

priority_queue<pii> pq;
vector<pii> adj[MAX_N];

int result[MAX_N];
int n, m, x;    //n = # of node, m = # of edge, x = end
int from, to , weight;
int ans = -1;

int dijkstra(int start, int end)
{
    for(int i=0; i< n+1; i++)
        result[i] = INF;
    
    pq = priority_queue<pii>();
    
    result[start] = 0;
    pq.push({0, start});
    
    while(!pq.empty())
    {
        pii cur = pq.top(); pq.pop();
        
        if (result[cur.second] != -1 * cur.first)
            continue;;
        
        for (pii edge:adj[cur.second])
        {
            int next_node = edge.second;
            int next_result = edge.first + (-1*cur.first);
            
            if (next_result < result[next_node])
            {
                result[next_node] = next_result;
                pq.push({-1*result[next_node], next_node});
            }
        }
        
    }
    
    return result[end];
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> n >> m >> x;
    
    for(int i=0; i<m; i++)
    {
        cin >> from >> to >> weight;
        adj[from].push_back({weight, to});
    }
    
    for(int i=1; i<n+1; i++)
    {
        int temp = dijkstra(i, x) + dijkstra(x, i);
        ans = max(ans, temp);
    }
    
    cout << ans<<"\n";
    
    
    
    return 0;
}
